### ISSUE TITLE
Support BBC-Micro Style TTL C-Sync, R, G, B

### ISSUE THREAD

**Issue by sharpie7** [breakintoprogram/pico-mposite#3]
Thanks for an excellent project. I got a nice stable picture with the monochrome composite config at the first attempt.

Question: I am interested in making a simple pattern generator to test and debug Microvitec Cub monitors at the [National Museum of Computing](https://www.tnmoc.org/). That needs a signal that matches that from a BBC Micro, which is TTL level (1 bit) signals with Composite Sync, R, G, and B with PAL timing.

Looking at your work already it seems that should be fairly (!!) simple to add as a new colour option to the existing project. Is this something you would regard as potentially in-scope and be willing to collaborate on, or would you prefer if I worked on it solo?

One thing that isn't obvious to me from the code (and I haven't looked very hard) is the significance of all the bits in the 16 bit values you assign to identifiers like HSLO in cvideo.h. Can you explain what the bits do?

**Comment by breakintoprogram** [comment]
Hi, no worries and thanks!

It should be fairly trivial to do a test signal generator for TTL monitors. The only complicated bit would be the TTL levels; the Pico GPIOs are 3.3V.

If I were to do that, I'd probably modify the colour version and:

- Discard the AD724 circuit and resistor ladders
- Use 3 GPIO pins for R, G and B (1 bit)
- Only have one SYNC GPIO pin rather than two
- Run those GPIO pins through a level converter to up them to TTL levels

You *may* get away with no level shifter as 3.3V GPIO is above the TTL trigger threshold.

The HSLO, HSHI, VSLO and VSHI values are for the low and high levels of the horizontal and vertical sync. These are specific bit values to set the GPIOs for the sync levels. For the monochrome version, this is a voltage of 0V (LO) and 0.4V (HI). For the colour version, it toggles two GPIO pins connected to the HSYNC and VSYNC pins on the AD724.

This is a little outside of the scope of this project, but am more than happy to help if you fork this project and need assistance.

**Comment by sharpie7** [comment]
Thanks. That makes sense. I always forget that the Pi Pico is 3.3V, but it's not the end of the world to add a level shifter!

So, if I take the colour version as a starting point then:
```
    #define colour_base 0x00
    #define colour_max  0xFF
    #define HSLO        0x4200
    #define HSHI        0x4000
    #define VSLO        0x4100
    #define VSHI        0x4000
    #define BORD        0x8000
```

I assume that for HSLO and VSLO that bits 9 and 10 are the bits to go out to the SYNC GPIO pins, but why 0x4 at the start?

Similarly, why is BORD 0x8000 and not just 0?

**Comment by breakintoprogram** [comment]
The 0x4nnn and 0x8nnn prefixes are ignored by the GPIO (as they are outside of the pins range of the PIO. The state machines use 0x0000 to exit a loop in cvidoe_sync.pio (line 20), and trigger cvideo_data.pio to start outputting pixel data (line 25).

As 0x0000 is a special word to say 'fill this block with pixel data', I needed to prevent HSHI and VSHI from being 0x0000, so use 0x4nnn to do that.

BORD is another special case; 0x8nnn marks out the border area. If you look at function set_border (cvideo.c, line 219), when setting the border lhs and rhs in the hsync table, I AND with BORD (0x8000). If that bit is set, I set the border colour, making sure I OR it with BORD (line 223).
